### PR TITLE
[Enhancement] Support Async delete files from Recycle Bin

### PR DESCRIPTION
**Pull request by srlch**
## Why I'm doing:
Currently, delete data in a sync operation for shared data mode in CatalogBin. The delete operation may take a very long time with the object lock for CatalogBin which may stuck sometime.

## What I'm doing:
Make file deletion in async way for both shared data mode and shared nothing mode.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -494,15 +508,33 @@ void eraseTable(long currentTimeMs) {
...
+            } else if (future.isDone()) {
+                try {
+                    finished = future.get();
+                } catch (Exception e) {
+                    finished = false;
+                }
+
+                if (!finished) {
+                    // finish with error, re-submit in next round
+                    asyncDeleteForTables.remove(info);
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:526]: worth logging with the error, in case of endless retry.

Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:528]: fixed

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -553,7 +605,7 @@ protected synchronized void erasePartition(long currentTimeMs) {
                 if (currentEraseOpCnt >= MAX_ERASE_OPERATIONS_PER_CYCLE) {
                     break;
                 }
-            } else {
+            } else if (asyncDeleteForPartitions.get(partitionInfo) == null) /* treated as error if task is not running */ {
```
[the remark this answers is not in the document]
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: fixed

File: fe/fe-core/src/main/java/com/starrocks/lake/Utils.java [hunk 3]
```diff
@@ -221,4 +222,29 @@ public static Optional<Long> getWarehouseIdByNodeId(SystemInfoService systemInfo
 
         return Optional.of(node.getWarehouseId());
     }
+
+    public static boolean checkAllAsyncTaskFinished(List<CompletableFuture<Boolean>> asyncTaskReturn) {
+        boolean allFinished = false;
+        allFinished = asyncTaskReturn.stream().allMatch(future -> future != null && future.isDone());
+        return allFinished;
```
[the remark this answers is not in the document]
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/Utils.java]: remove this useless code.

**Review by kevincai — DISMISSED** [review]

**Review by kevincai — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**